sns_ik_lib/src/sns_vel_ik_base.cpp

```diff
@@ -0,0 +1,393 @@
+  if (dx.size() != J.rows()) { ROS_ERROR("Bad Input!  dx.size() != J.rows()"); return false; }
+  if (J.cols() != dqNull.rows()) { ROS_ERROR("Bad Input!  J.cols() != dqNull.rows()"); return false; }
+  if (!dq) { ROS_ERROR("Bad Input!  dq is nullptr!"); return false; }
+  if (!resErr) { ROS_ERROR("Bad Input!  resErr is nullptr!"); return false; }
+
+  // Solve the linear system
+  Eigen::MatrixXd B = dx - J*dqNull;
+  Eigen::MatrixXd X = linSolver_.solve(B);
+  if(linSolver_.info() != Eigen::ComputationInfo::Success) {
+    ROS_ERROR("Failed to solve projection equations!");
+    return false;
+  }
+
+  // Solve for dq
+  *dq = X + dqNull;
+  *resErr = (JW*X - B).squaredNorm();
+  return true;
+}
+
+/*************************************************************************************************/
+
+SnsIkExitCode SnsVelIkBase::computeTaskScalingFactor(const Eigen::MatrixXd& J, const Eigen::MatrixXd& JW,
+                                            const Eigen::VectorXd& dx, const Eigen::VectorXd& dq,
+                                            const std::vector<bool>& jntIsFree,
+                                            double* taskScale, int* jntIdx, double* resErr)
+{
+  if (J.cols() != JW.cols()) { ROS_ERROR("Bad Input!  J.cols() != JW.cols()"); return SnsIkExitCode::BadUserInput; }
+  if (J.rows() != JW.rows()) { ROS_ERROR("Bad Input!  J.rows() != JW.rows()"); return SnsIkExitCode::BadUserInput; }
+  if (dx.size() != J.rows()) { ROS_ERROR("Bad Input!  dx.size() != J.rows()"); return SnsIkExitCode::BadUserInput; }
+  if (dq.size() != J.cols()) { ROS_ERROR("Bad Input!  dq.size() != J.cols()"); return SnsIkExitCode::BadUserInput; }
+  if (!taskScale) { ROS_ERROR("taskScale is nullptr!"); return SnsIkExitCode::BadUserInput; }
+  if (!jntIdx) { ROS_ERROR("jntIdx is nullptr!"); return SnsIkExitCode::BadUserInput; }
+  if (!resErr) { ROS_ERROR("resErr is nullptr!"); return SnsIkExitCode::BadUserInput; }
+
+  // Compute "a" and "b" from the paper.   (J*W*a = dx)
+  Eigen::VectorXd a = linSolver_.solve(dx);
+  if(linSolver_.info() != Eigen::ComputationInfo::Success) {
+    ROS_ERROR("Failed to solve projection equations!");
+    return SnsIkExitCode::InternalError;
+  }
+  *resErr = (JW*a - dx).squaredNorm();
+  Eigen::ArrayXd b = (dq - a).array();
+
+  // Compute the task scale associated with each joint
+  Eigen::ArrayXd jntScaleFactorArr(nJnt_);
+  Eigen::ArrayXd lowMargin = (dqLow_ - b);
+  Eigen::ArrayXd uppMargin = (dqUpp_ - b);
+  for (int i = 0; i < nJnt_; i++) {
+    if (jntIsFree[i]) {
+      jntScaleFactorArr(i) = fingScaleFactor(lowMargin(i), uppMargin(i), a(i));
+    } else {  // joint is constrained
+      jntScaleFactorArr(i) = POS_INF;
+    }
+  }
+
+  // Compute the most critical scale factor and corresponding joint index
+  *jntIdx = 0;  // index of the most critical joint
+  *taskScale = jntScaleFactorArr(0);  // minimum value of jntScaleFactorArr()
```

Comment:
I got lost in figuring out why we initializing it with the first element of `jntScaleFactorArr`?

Reply:
Eigen does not have a nice method to compute both the minimum value and the minimum index, so this loop does the task. Rather than putting a special case inside of the loop to initialize the "best so far" data, I initialize it before the loop and then start the loop on element one. Perhaps I should just make this into a function.